### PR TITLE
merge develop-controller to master

### PR DESCRIPTION
**Pull request by alwayswannajava**

### CODE REVIEW

File: src/main/java/com/spring/onlinebookstore/controller/BookController.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/controller/BookController.java:16]:
```suggestion
@RequestMapping("/books")
@RequiredArgsConstructor
```

File: src/main/java/com/spring/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+public class BookController {
+    private final BookService bookService;
+
+    @Autowired
+    public BookController(BookService bookService) {
+        this.bookService = bookService;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/controller/BookController.java]:
```suggestion
```

File: src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.spring.onlinebookstore.dto;
+
+import jakarta.annotation.Nonnull;
+import jakarta.persistence.Column;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class CreateBookRequestDto {
+    @Nonnull
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    private String title;
+    @Nonnull
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    private String author;
+    @Column(unique = true)
+    @Nonnull
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    private String isbn;
+    @Nonnull
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotNull
```

File: src/main/java/com/spring/onlinebookstore/model/Book.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@Table(name = "books")
+@Getter
+@Setter
+@NoArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/model/Book.java]:
```suggestion
```

File: src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import com.spring.onlinebookstore.model.Book;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.EntityTransaction;
+import java.util.List;
+import java.util.Optional;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java:13]:
```suggestion
@Repository
@RequiredArgsConstructor
```

File: src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+public class BookRepositoryImpl implements BookRepository {
+    private final EntityManagerFactory entityManagerFactory;
+
+    @Autowired
+    public BookRepositoryImpl(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java]:
```suggestion
```

File: src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        return book;
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery("from Book", Book.class)
+                    .getResultList();
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java:48]: add catch

File: src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public Optional<Book> findById(int id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return Optional.ofNullable(entityManager.find(Book.class, id));
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/repository/BookRepositoryImpl.java:57]: add catch

File: src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import com.spring.onlinebookstore.dto.CreateBookRequestDto;
+import com.spring.onlinebookstore.exception.EntityNotFoundException;
+import com.spring.onlinebookstore.mapper.BookMapper;
+import com.spring.onlinebookstore.model.Book;
+import com.spring.onlinebookstore.repository.BookRepository;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java:13]:
```suggestion
@Service
@RequiredArgsConstructor
```

File: src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+public class BookServiceImpl implements BookService {
+    private final BookRepository bookRepository;
+    private final BookMapper bookMapper;
+
+    @Autowired
+    public BookServiceImpl(BookRepository bookRepository, BookMapper bookMapper) {
+        this.bookRepository = bookRepository;
+        this.bookMapper = bookMapper;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/spring/onlinebookstore/controller/BookController.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    @GetMapping("/{id}")
+    public BookDto getBookById(@PathVariable int id) {
+        return bookService.findById(id);
+    }
+
+    @PostMapping
+    public BookDto addBook(@RequestBody CreateBookRequestDto bookRequestDto) {
+        return bookService.save(bookRequestDto);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/spring/onlinebookstore/controller/BookController.java]:
```suggestion
```

File: src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import java.math.BigDecimal;
+
+public record CreateBookRequestDto(
+        @NotBlank
+        String title,
+        @NotBlank
+        String author,
+        @Column(nullable = false, unique = true)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/spring/onlinebookstore/dto/CreateBookRequestDto.java]: Use Column annotation only with fields of entity class. Also, you can validate ISBN format using annotation Pattern and some regex from google) 
```suggestion
        @NotBlank
```

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Everything good but one important issue should be fixed 😉

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**